File: src/main/java/aurocosh/divinefavor/common/item/talismans/spell/SpellTalismanArrowThrow.java
package aurocosh.divinefavor.common.item.talismans.spell;

import aurocosh.divinefavor.common.item.talismans.spell.base.ItemSpellTalisman;
import aurocosh.divinefavor.common.item.talismans.spell.base.SpellOptions;
import aurocosh.divinefavor.common.item.talismans.spell.base.TalismanContext;
import aurocosh.divinefavor.common.spirit.base.ModSpirit;
import aurocosh.divinefavor.common.util.UtilRandom;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.projectile.EntityArrow;
import net.minecraft.init.Items;
import net.minecraft.init.SoundEvents;
import net.minecraft.item.ItemArrow;
import net.minecraft.item.ItemStack;
import net.minecraft.util.EnumHand;
import net.minecraft.util.SoundCategory;
import net.minecraft.world.World;

import java.util.EnumSet;

public class SpellTalismanArrowThrow extends ItemSpellTalisman {
    public SpellTalismanArrowThrow(String name, ModSpirit spirit, int favorCost, EnumSet<SpellOptions> options) {
        super(name, spirit, favorCost, options);
    }

    @Override
    protected void performActionServer(TalismanContext context) {
        ItemStack stack = context.player.getHeldItem(context.hand);
        shootArrow(stack, context.world, context.player, 20000);
    }

    /**
     * Gets the velocity of the arrow entity from the bow's favorCost
     */
    public static float getArrowVelocity(int charge) {
        float f = (float) charge / 20.0F;
        f = (f * f + f * 2.0F) / 3.0F;
        return f < 1.0F ? f : 1.0F;
    }

    private ItemStack findAmmo(EntityPlayer player) {
        if (this.isArrow(player.getHeldItem(EnumHand.OFF_HAND)))
            return player.getHeldItem(EnumHand.OFF_HAND);
        else if (this.isArrow(player.getHeldItem(EnumHand.MAIN_HAND)))
            return player.getHeldItem(EnumHand.MAIN_HAND);
        else {
            for (int i = 0; i < player.inventory.getSizeInventory(); ++i) {
                ItemStack itemstack = player.inventory.getStackInSlot(i);

                if (this.isArrow(itemstack))
                    return itemstack;
            }

            return ItemStack.EMPTY;
        }
    }

    protected boolean isArrow(ItemStack stack) {
        return stack.getItem() instanceof ItemArrow;
    }

    /**
     * Called when the player stops using an Item (stops holding the right mouse button).
     */
    public void shootArrow(ItemStack stack, World worldIn, EntityLivingBase entityLiving, int charge) {
        if (entityLiving instanceof EntityPlayer) {
            EntityPlayer entityplayer = (EntityPlayer) entityLiving;
            boolean flag = entityplayer.capabilities.isCreativeMode;
            ItemStack itemstack = this.findAmmo(entityplayer);

            charge = net.minecraftforge.event.ForgeEventFactory.onArrowLoose(stack, worldIn, entityplayer, charge, !itemstack.isEmpty() || flag);
            if (charge < 0)
                return;

            if (!itemstack.isEmpty() || flag) {
                if (itemstack.isEmpty())
                    itemstack = new ItemStack(Items.ARROW);

                float f = getArrowVelocity(charge);

                if ((double) f >= 0.1D) {
                    boolean flag1 = entityplayer.capabilities.isCreativeMode || (itemstack.getItem() instanceof ItemArrow && ((ItemArrow) itemstack.getItem()).isInfinite(itemstack, stack, entityplayer));

                    if (!worldIn.isRemote) {
                        ItemArrow itemarrow = (ItemArrow) (itemstack.getItem() instanceof ItemArrow ? itemstack.getItem() : Items.ARROW);
                        EntityArrow entityarrow = itemarrow.createArrow(worldIn, itemstack, entityplayer);
                        entityarrow.shoot(entityplayer, entityplayer.rotationPitch, entityplayer.rotationYaw, 0.0F, f * 3.0F, 1.0F);

                        if (f == 1.0F)
                            entityarrow.setIsCritical(true);

                        stack.damageItem(1, entityplayer);

                        if (flag1 || entityplayer.capabilities.isCreativeMode && (itemstack.getItem() == Items.SPECTRAL_ARROW || itemstack.getItem() == Items.TIPPED_ARROW))
                            entityarrow.pickupStatus = EntityArrow.PickupStatus.CREATIVE_ONLY;

                        worldIn.spawnEntity(entityarrow);
                    }

                    worldIn.playSound(null, entityplayer.posX, entityplayer.posY, entityplayer.posZ, SoundEvents.ENTITY_ARROW_SHOOT, SoundCategory.PLAYERS, 1.0F, 1.0F / (UtilRandom.random.nextFloat() * 0.4F + 1.2F) + f * 0.5F);

                    if (!flag1 && !entityplayer.capabilities.isCreativeMode) {
                        itemstack.shrink(1);

                        if (itemstack.isEmpty())
                            entityplayer.inventory.deleteStack(itemstack);
                    }
                }
            }
        }
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/item/spell_talismans/SpellTalismanArrowThrow.kt
package aurocosh.divinefavor.common.item.spell_talismans

import aurocosh.divinefavor.common.item.spell_talismans.base.ItemSpellTalisman
import aurocosh.divinefavor.common.item.spell_talismans.base.SpellOptions
import aurocosh.divinefavor.common.item.spell_talismans.base.TalismanContext
import aurocosh.divinefavor.common.spirit.base.ModSpirit
import aurocosh.divinefavor.common.util.UtilPlayer
import aurocosh.divinefavor.common.util.UtilRandom
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.entity.projectile.EntityArrow
import net.minecraft.init.Items
import net.minecraft.init.SoundEvents
import net.minecraft.item.ItemArrow
import net.minecraft.item.ItemStack
import net.minecraft.util.SoundCategory
import net.minecraft.world.World
import net.minecraftforge.event.ForgeEventFactory
import java.util.*

class SpellTalismanArrowThrow(name: String, spirit: ModSpirit, favorCost: Int, options: EnumSet<SpellOptions>) : ItemSpellTalisman(name, spirit, favorCost, options) {

    override fun performActionServer(context: TalismanContext) {
        val stack = context.player.getHeldItem(context.hand)
        shootArrow(stack, context.world, context.player, 15)
    }

    private fun findAmmo(player: EntityPlayer): ItemStack {
        val stack = UtilPlayer.getItemInHand(player) { it is ItemArrow }
        if (!stack.isEmpty)
            return stack

        for (i in 0 until player.inventory.sizeInventory) {
            val itemStack = player.inventory.getStackInSlot(i)
            if (itemStack.item is ItemArrow)
                return itemStack
        }

        return ItemStack.EMPTY
    }

    private fun shootArrow(stack: ItemStack, worldIn: World, entityLiving: EntityLivingBase, charge: Int) {
        var chargeCounter = charge
        if (entityLiving !is EntityPlayer)
            return
        val flag = entityLiving.capabilities.isCreativeMode
        var itemStack = this.findAmmo(entityLiving)

        chargeCounter = ForgeEventFactory.onArrowLoose(stack, worldIn, entityLiving, chargeCounter, !itemStack.isEmpty || flag)
        if (chargeCounter < 0)
            return

        if (!itemStack.isEmpty || flag) {
            if (itemStack.isEmpty)
                itemStack = ItemStack(Items.ARROW)

            val f = getArrowVelocity(chargeCounter)

            if (f.toDouble() >= 0.1) {
                val flag1 = entityLiving.capabilities.isCreativeMode || itemStack.item is ItemArrow && (itemStack.item as ItemArrow).isInfinite(itemStack, stack, entityLiving)

                if (!worldIn.isRemote) {
                    val itemarrow = (if (itemStack.item is ItemArrow) itemStack.item else Items.ARROW) as ItemArrow
                    val entityarrow = itemarrow.createArrow(worldIn, itemStack, entityLiving)
                    entityarrow.shoot(entityLiving, entityLiving.rotationPitch, entityLiving.rotationYaw, 0.0f, f * 3.0f, 1.0f)

                    if (f == 1.0f)
                        entityarrow.isCritical = true

                    stack.damageItem(1, entityLiving)

                    if (flag1 || entityLiving.capabilities.isCreativeMode && (itemStack.item === Items.SPECTRAL_ARROW || itemStack.item === Items.TIPPED_ARROW))
                        entityarrow.pickupStatus = EntityArrow.PickupStatus.CREATIVE_ONLY

                    worldIn.spawnEntity(entityarrow)
                }

                worldIn.playSound(null, entityLiving.posX, entityLiving.posY, entityLiving.posZ, SoundEvents.ENTITY_ARROW_SHOOT, SoundCategory.PLAYERS, 1.0f, 1.0f / (UtilRandom.random.nextFloat() * 0.4f + 1.2f) + f * 0.5f)

                if (!flag1 && !entityLiving.capabilities.isCreativeMode) {
                    itemStack.shrink(1)

                    if (itemStack.isEmpty)
                        entityLiving.inventory.deleteStack(itemStack)
                }
            }
        }
    }

    companion object {
        fun getArrowVelocity(charge: Int): Float {
            var f = charge.toFloat() / 20.0f
            f = (f * f + f * 2.0f) / 3.0f
            return if (f < 1.0f) f else 1.0f
        }
    }
}
